synapse/replication/tcp/external_cache.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import TYPE_CHECKING, Any, Optional
+
+from prometheus_client import Counter
+
+from synapse.util import json_decoder, json_encoder
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+set_counter = Counter(
+    "synapse_external_cache_set",
+    "Number of times we set a cache",
+    labelnames=["cache_name"],
+)
+
+get_counter = Counter(
+    "synapse_external_cache_get",
+    "Number of times we get a cache",
+    labelnames=["cache_name", "hit"],
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalCache:
+    """A cache backed by an external Redis. Does nothing if no Redis is
+    configured.
+    """
+
+    def __init__(self, hs: "HomeServer"):
+        self._redis_connection = None
+
+        if hs.config.redis.redis_enabled:
+            from synapse.replication.tcp.redis import lazyConnection
+
+            logger.info(
+                "Connecting to redis (host=%r port=%r) for external cache",
+                hs.config.redis_host,
+                hs.config.redis_port,
+            )
+
+            # First create the connection for sending commands.
+            self._redis_connection = lazyConnection(
```

Comment:
Yeah, that's a good point. I'm not really sure how a) cheap the connections are and b) whether we want to segregate the cache stuff from the replication pub/sub. Since the commands get pipelined the connections shouldn't get congested, I guess?

Reply:
I've changed it so that we share the outbound connection.